replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,259 @@
+
+  private final NodeAdapter local;
+
+  private final NodeAdapter destination;
+
+  private final Queue queue;
+
+  /**
+   * Creates a new worker which will monitor the provided queue and perform operations between the
+   * provided local and destination sites.
+   *
+   * Does not begin monitoring of the queue until {@link #run()} is called.
+   *
+   * @param queue the queue to pull tasks from; corresponding queue for the destination site
+   * @param local the local site adapter
+   * @param destination the destination site adapter
+   */
+  public Worker(Queue queue, NodeAdapter local, NodeAdapter destination) {
+    this.queue = queue;
+    this.local = local;
+    this.destination = destination;
+  }
+
+  /**
+   * Begin monitoring a queue for tasks. The handler of this worker is responsible for halting this
+   * thread by calling interrupt on this worker's thread, finishing the transfer of the current task
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    while (running || !Thread.currentThread().isInterrupted()) {
+      try {
+        Task task = queue.take();
+        this.processing.set(true);
+        if (supported(task)) {
+          this.handle(task);
+        }
+        this.processing.set(false);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Set the running flag to false and wait for the thread to die. If an item is currently in the
+   * process of transferring, it will finish.
+   */
+  public void cancel() {
+    this.running = false;
+    try {
+      Thread.currentThread().join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * @return {@code true} if this worker is transferring an item, otherwise {@code false}
+   */
+  public boolean isProcessing() {
+    return processing.get();
+  }
+
+  private boolean supported(Task task) throws InterruptedException {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          String.format("Operation task %s is not supported by this worker", task));
+      return false;
+    }
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    boolean proceed = checkAdapterAvailabilities(task);
+    if (!proceed) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(t -> t instanceof DDFMetadataInfo)
+        .map(DDFMetadataInfo.class::cast)
+        .collect(Collectors.toList());
+
+    if (infos.size() > 1) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "more than one metadata is currently not supported by this worker");
+      return;
+    }
+
+    if(infos.isEmpty()) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "unable to handle any of the metadatas received in this task");
+      return;
+    }
+
+    final DDFMetadataInfo info = infos.get(0);
```

Comment:
❓ Should this worker be more generic and not care about DDF metadata info but instead pass that info to the adapter and let the adapter handle it properly. A ddf adapter would care and on a query, it would simply return it back whereas as another one would perform its query to get the data.

Reply:
Yeah I think eventually we will want that, but I'd like to avoid modifications to the adapter API in this PR if we can defer it.